django/contrib/admin/options.py

```diff
@@ -1320,13 +1320,9 @@ def response_change(self, request, obj):
             self.message_user(request, msg, messages.SUCCESS)
             return self.response_post_save_change(request, obj)
 
-    def response_post_save_add(self, request, obj):
-        """
-        Figure out where to redirect after the 'Save' button has been pressed
-        when adding a new object.
-        """
+    def _response_post_save(self, request, obj):
         opts = self.model._meta
-        if self.has_change_permission(request, None):
+        if self.has_view_permission(request) or self.has_change_permission(request):
```

Comment:
Unless I am missing something here, you only need `self.has_view_permission(request)`, since it checks for view permissions or change permission.

```
 def has_view_permission(self, request, obj=None):
     """
     Return True if the given request has permission to view the given
     Django model instance. The default implementation doesn't examine the
     `obj` parameter.
 
     If overridden by the user in subclasses, it should return True if the
     given request has permission to view the `obj` model instance. If `obj`
     is None, it should return True if the request has permission to view
     any object of the given type.
     """
     opts = self.opts
     codename_view = get_permission_codename('view', opts)
     codename_change = get_permission_codename('change', opts)
     return (
         request.user.has_perm('%s.%s' % (opts.app_label, codename_view)) or
         request.user.has_perm('%s.%s' % (opts.app_label, codename_change))
)
```

Reply:
Thanks for the review. That simplification is appealing however there are lots of examples in `options.py` that don't make this simplification:

```
./django/contrib/admin/options.py:591:                if not (inline.has_view_permission(request, obj) or
./django/contrib/admin/options.py:1325:        if self.has_view_permission(request) or self.has_change_permission(request):
./django/contrib/admin/options.py:1551:            if not self.has_view_permission(request, obj) and not self.has_change_permission(request, obj):
./django/contrib/admin/options.py:1673:        if not self.has_view_permission(request) and not self.has_change_permission(request):
./django/contrib/admin/options.py:1899:        if not self.has_view_permission(request, obj) and not self.has_change_permission(request, obj):
./django/contrib/admin/options.py:2106:        if not self.has_change_permission(request) and not self.has_view_permission(request):
```

So I think it would be inconsistent to do so here. If a project were to override `has_view_permission` it might not include `has_change_permission` and then the assumption that allows this simplification would not be valid. Therefore, I think it would be best to follow the existing pattern and keep both checks.

Reply:
Sounds good to me.